lib/VolSurface/Calibration/Equities.pm

```diff
@@ -0,0 +1,648 @@
+package VolSurface::Calibration::Equities;
+
+use strict;
+use warnings;
+
+our $VERSION = '0.01';
+
+=head1 NAME
+
+VolSurface::Calibration::Equities
+
+=head1 DESCRIPTION
+
+This module implements Binary.com's volatility surface calibration algorithm which is based on SABR.
+The SABR (Stochastic alpha, beta, rho) model is a stochastic volatility model, which is used to 
+estimate the volatility smile in the derivatives market. For a more general overview of the general 
+SABR model, please refer https://en.wikipedia.org/wiki/SABR_volatility_model.
+
+=head1 VERSION
+
+Version 0.01
+
+=cut
+
+use Moose;
+
+use Date::Utility;
+use List::MoreUtils qw(first_index);
+use List::Util qw(min max);
+use Math::Trig qw(tanh);
+use Try::Tiny;
+use Format::Util::Numbers qw(roundnear);
+
+=head2 calibration_param_names
+
+Calibration parameter names.
+It is hard-coded here because it needs to be in this sequence.
+
+=cut
+
+my @calibration_param_names = qw/
+    atmvolshort
+    atmvol1year
+    atmvolLong
+    atmWingL
+    atmWingR
+    skewshort
+    skew1year
+    skewlong
+    skewwingL
+    skewwingR
+    kurtosisshort
+    kurtosislong
+    kurtosisgrowth/;
+
+has surface => (
+    is       => 'ro',
+    isa      => 'HashRef',
+    required => 1,
+);
+
+=head2 term_by_day
+
+Get all the terms in a surface in ascending order.
+
+=cut
+
+has term_by_day => (
+    is       => 'ro',
+    isa      => 'ArrayRef',
+    required => 1,
+);
+
+=head2 parameterization
+
+The parameterized (and, thus, smoothed) version of this surface.
+
+=cut
+
+has parameterization => (
+    is  => 'rw',
+    isa => 'Maybe[HashRef]',
+);
+
+=head2 smile_points
+
+The points across a smile.
+
+It can be delta points, moneyness points or any other points that we might have in the future.
+
+=cut
+
+has smile_points => (
+    is       => 'ro',
+    isa      => 'ArrayRef',
+    required => 1,
+);
+
+has _max_iteration => (
+    is      => 'ro',
+    isa     => 'Int',
+    default => 500,
+);
+
+has _tolerance_level => (
+    is      => 'ro',
+    isa     => 'Num',
+    default => 1e-05,
+);
+
+has _scale => (
+    is      => 'ro',
+    isa     => 'Num',
+    default => 10,
+);
+
+=head2 function_to_optimize
+
+The function that we want to optimize.
+
+=cut
+
+sub function_to_optimize {
+    my ($self, $params) = @_;
+
+    # this number is returned when the optimization is heading the wrong path
+    my $error = 1000;
+
+    my %params;
+    @params{@calibration_param_names} = @$params;
+
+    my @tenors = @{$self->term_by_day};
+    my @variance =
+        map { $self->_calculate_variance($_ / 365, \%params) } @tenors;
+
+    foreach my $param_name (@calibration_param_names) {
+        my $param_value = $params{$param_name};
+        return $error
+            if ($param_name =~ /wing(R|L)/i && max(abs($param_value)) > 20.0);
```

Comment:
why `max()`?

The return value of `abs()` is a single scalar. So, you are looking for the maximum in a set of 1 element. That's always the element itself.